### PR TITLE
bpo-46841: Quicken code in-place

### PR DESCRIPTION
**Pull request by brandtbucher**
This moves the bytecode to the end of the corresponding `PyCodeObject`, and quickens it in-place.

Related changes:
- `PyCodeObject` is now more compact. I've removed the almost-always-unused `co_varnames`, `co_freevars`, and `co_cellvars` member caches, and rearranged some `int` members to fill some holes in the struct on 64-bit builds. `co_code` has been removed and replaced with `_PyCode_GetCode`, and `co_quickened` and `co_firstinstr` have been replaced with `_PyCode_CODE`.
- `_PyOpcode_Deopt` is a new mapping from all opcodes to their un-quickened forms.
- `_PyOpcode_InlineCacheEntries` is renamed to `_PyOpcode_Caches`, `_Py_IncrementCountAndMaybeQuicken` is renamed to `_PyCode_Warmup`, `_Py_Quicken` is renamed to `_PyCode_Quicken`, and `_co_quickened` is renamed to `_co_code_adaptive` (and is now a read-only `memoryview`).
- We don't emit unused nonzero opargs anymore in the compiler.

https://bugs.python.org/issue46841

### CODE REVIEW

File: Lib/opcode.py [hunk 1]
```diff
@@ -229,77 +229,111 @@ def jabs_op(name, op, entries=0):
...
+        'BINARY_OP_MULTIPLY_INT',
+        'BINARY_OP_SUBTRACT_FLOAT',
+        'BINARY_OP_SUBTRACT_INT',
+    ],
+    'BINARY_SUBSCR': [
+        'BINARY_SUBSCR_ADAPTIVE',
+        'BINARY_SUBSCR_DICT',
+        'BINARY_SUBSCR_GETITEM',
+        'BINARY_SUBSCR_LIST_INT',
+         'BINARY_SUBSCR_TUPLE_INT',
```
Comment (markshannon) [inline Lib/opcode.py]: alignment

File: Objects/codeobject.c [hunk 2]
```diff
@@ -341,16 +338,14 @@ init_code(PyCodeObject *co, struct _PyCodeConstructor *con)
...
-    co->co_cellvars = NULL;
-    co->co_freevars = NULL;
 
     /* not set */
     co->co_weakreflist = NULL;
     co->co_extra = NULL;
 
     co->co_warmup = QUICKENING_INITIAL_WARMUP_VALUE;
-    co->co_quickened = NULL;
+    memcpy(_PyCode_CODE(co), PyBytes_AS_STRING(con->code),
```
Comment (markshannon) [inline Objects/codeobject.c:347]: This feels inefficient. When unmarshalling, we create a `bytes` object, just to copy its contents and destroy it again.

Probably best done in another PR, but could you add it to the issue.

File: Objects/codeobject.c [hunk 3]
```diff
@@ -1151,46 +1140,39 @@ _PyCode_SetExtra(PyObject *code, Py_ssize_t index, void *extra)
 PyObject *
 _PyCode_GetVarnames(PyCodeObject *co)
```
Comment (markshannon) [inline Objects/codeobject.c:1141]: This no longer gets the `varnames`, it makes it.
Also, why this change?

File: Objects/codeobject.c [hunk 4]
```diff
@@ -1538,42 +1521,41 @@ code_getfreevars(PyCodeObject *code, void *closure)
 }
 
 static PyObject *
-code_getquickened(PyCodeObject *code, void *closure)
+code_getundercode(PyCodeObject *code, void *closure)
```
Comment (markshannon) [inline Objects/codeobject.c]: "undercode"? underlying_code, presumably?
I think quickened is fine, or maybe "adaptive", as it isn't necessarily quickened yet.

File: Objects/codeobject.c [hunk 5]
```diff
@@ -1538,42 +1521,41 @@ code_getfreevars(PyCodeObject *code, void *closure)
...
+static PyObject *
+code_getcode(PyCodeObject *code, void *closure)
+{
+    return _PyCode_GetCode(code);
 }
 
 static PyGetSetDef code_getsetlist[] = {
-    {"co_lnotab",    (getter)code_getlnotab, NULL, NULL},
+    {"co_lnotab",   (getter)code_getlnotab,    NULL, NULL},
+    {"_co_code",    (getter)code_getundercode, NULL, NULL},
```
Comment (markshannon) [inline Objects/codeobject.c]: :+1: This will be handy.

The name is too close to `co_code`, IMO.
I'd prefer `_co_adaptive_code`, the longer name is also a clue that this is more complex than just the bytecode.

File: Objects/frameobject.c [hunk 6]
```diff
@@ -493,7 +499,7 @@ frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignore
 
     /* PyCode_NewWithPosOnlyArgs limits co_code to be under INT_MAX so this
      * should never overflow. */
-    int len = (int)(PyBytes_GET_SIZE(f->f_frame->f_code->co_code) / sizeof(_Py_CODEUNIT));
+    int len = Py_SIZE(f->f_frame->f_code);
```
Comment (markshannon) [inline Objects/frameobject.c]: MSVC warning. You need the `(int)` cast.

File: Objects/frameobject.c [hunk 7]
```diff
@@ -838,12 +844,23 @@ PyFrame_New(PyThreadState *tstate, PyCodeObject *code,
 static int
 _PyFrame_OpAlreadyRan(_PyInterpreterFrame *frame, int opcode, int oparg)
```
Comment (markshannon) [inline Objects/frameobject.c:845]: We might be able to remove this entirely, as there should be no way to trace before the first `RESUME`.
So don't worry too much if it appears a bit broken.

File: Objects/frameobject.c [hunk 8]
```diff
@@ -862,7 +879,10 @@ _PyFrame_FastToLocalsWithError(_PyInterpreterFrame *frame) {
     }
     co = frame->f_code;
     fast = _PyFrame_GetLocalsArray(frame);
-    if (frame->f_lasti < 0 && _Py_OPCODE(co->co_firstinstr[0]) == COPY_FREE_VARS) {
+    // COPY_FREE_VARS has no quickened forms, so no need to use _PyOpcode_Deopt
```
Comment (markshannon) [inline Objects/frameobject.c:882]: We should (in theory) be able to assert  `frame->f_lasti >= 0` here.

File: Python/ceval.c [hunk 9]
```diff
@@ -1566,10 +1565,17 @@ trace_function_exit(PyThreadState *tstate, _PyInterpreterFrame *frame, PyObject
...
 {
-    _Py_CODEUNIT *instrs = (_Py_CODEUNIT *)PyBytes_AS_STRING(code->co_code);
-    while (offset > 0 && _Py_OPCODE(instrs[offset-1]) == EXTENDED_ARG) {
+    // You typically cannot scan backwards like this over quickened code, since
+    // inline cache entries might *appear* to be valid instructions. However,
+    // our check for oparg makes this particular case safe: the instruction at
+    // the current offset can *only* be an EXTENDED_ARG iff oparg is still
+    // nonzero. Also, EXTENDED_ARG has no quickened forms, so no need to use
+    // _PyOpcode_Deopt here:
+    while (oparg && _Py_OPCODE(_PyCode_CODE(code)[offset]) == EXTENDED_ARG) {
```
Comment (markshannon) [inline Python/ceval.c]: Simply removing `skip_backwards_over_extended_args` and replacing
`int instr_prev = skip_backwards_over_extended_args(frame->f_code, frame->f_lasti);` 
with `int instr_prev = frame->f_lasti;` passes all the tests on main.

File: Python/ceval.c [hunk 10]
```diff
@@ -1660,10 +1666,15 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
         consts = co->co_consts; \
-        first_instr = co->co_firstinstr; \
+        first_instr = _PyCode_CODE(co); \
     } \
     assert(frame->f_lasti >= -1); \
-    next_instr = first_instr + frame->f_lasti + 1; \
+    /* Jump back to the last instruction executed... */ \
+    next_instr = first_instr + frame->f_lasti; \
+    /* ...past any inline cache entries... */ \
+    next_instr += _PyOpcode_Caches[_PyOpcode_Deopt[_Py_OPCODE(*next_instr)]]; \
```
Comment (markshannon) [inline Python/ceval.c]: I don't like this; it adds a chain of dependent loads to what is a fairly hot path in the interpreter.

Rather than attempting to match the behavior to the name, it might be better to match the name to the behavior.
We could rename `f_lasti` to `f_nexti_less_one` or something similar.


Now that the first instruction is at a fixed offset from the start of the code object, we will want to store `next_instr` instead of the offset, replacing `f->f_lasti = INSTR_OFFSET();` with `f->f_next_instr = next_instr` (in another PR).

File: Objects/codeobject.c [hunk 11]
```diff
@@ -1151,46 +1140,39 @@ _PyCode_SetExtra(PyObject *code, Py_ssize_t index, void *extra)
...
+        int opcode = _PyOpcode_Deopt[_Py_OPCODE(instruction)];
+        int caches = _PyOpcode_Caches[opcode];
+        instructions[i] = _Py_MAKECODEUNIT(opcode, _Py_OPARG(instruction));
+        while (caches--) {
+            instructions[++i] = _Py_MAKECODEUNIT(CACHE, 0);
         }
     }
-    Py_INCREF(co->co_freevars);
-    return co->co_freevars;
+    return code;
```
Comment (markshannon) [inline Objects/codeobject.c:1175]: I think we want to cache this.
It is relatively expensive to create and we will need to cache it for bytecode instrumentation (which loses the mapping to original bytecode).

**Review by markshannon — COMMENTED** [review]
Generally good, but I'm a little concerned about the cost of finding `next_instr` when resuming a frame.

https://github.com/python/cpython/pull/31901 should simplify things a bit, regarding scanning backwards.

File: Objects/codeobject.c (hunk 2 above)
Comment (brandtbucher) [reply Objects/codeobject.c:347]: Yeah, I originally changed the `marshal` format a bit in this branch, but I decided to leave that for later to reduce the scope of this PR.

There are other places in `marshal` where we make unnecessary intermediate copies when reading. Those can maybe be cleaned up at the same time.

File: Python/ceval.c (hunk 10 above)
Comment (brandtbucher) [reply Python/ceval.c]: I'll just revert this change, and leave the name as-is for now. The name isn't exactly *wrong*, if you view `CACHE` entries as instructions.

I've added a task to the issue to add `f_next_instr` later.

**Comment by brandtbucher** [comment]
It looks like this results in a 3% memory improvement across all benchmarks:

<details>

```
Slower (1):
- regex_dna: 13.9 MB +- 374.4 kB -> 14.8 MB +- 311.7 kB: 1.07x slower

Faster (49):
- xml_etree_generate: 12.8 MB +- 340.8 kB -> 11.5 MB +- 329.9 kB: 1.11x faster
- logging_simple: 16.4 MB +- 1428.4 kB -> 14.9 MB +- 1906.2 kB: 1.10x faster
- html5lib: 27.4 MB +- 1000.0 kB -> 25.6 MB +- 1662.1 kB: 1.07x faster
- xml_etree_process: 12.8 MB +- 467.9 kB -> 12.1 MB +- 371.3 kB: 1.06x faster
- pidigits: 7594.4 kB +- 233.7 kB -> 7238.7 kB +- 100.3 kB: 1.05x faster
- regex_compile: 8885.3 kB +- 424.5 kB -> 8524.6 kB +- 159.0 kB: 1.04x faster
- unpickle: 8016.9 kB +- 384.9 kB -> 7694.8 kB +- 178.5 kB: 1.04x faster
- telco: 8009.3 kB +- 358.4 kB -> 7690.8 kB +- 327.4 kB: 1.04x faster
- pickle_pure_python: 8018.5 kB +- 390.8 kB -> 7706.8 kB +- 231.7 kB: 1.04x faster
- json_loads: 7637.5 kB +- 239.3 kB -> 7350.9 kB +- 186.3 kB: 1.04x faster
- chaos: 8442.3 kB +- 116.6 kB -> 8125.8 kB +- 39.7 kB: 1.04x faster
- logging_silent: 8057.6 kB +- 264.3 kB -> 7761.4 kB +- 169.1 kB: 1.04x faster
- fannkuch: 7315.6 kB +- 213.7 kB -> 7061.8 kB +- 186.3 kB: 1.04x faster
- pickle_dict: 7948.3 kB +- 257.2 kB -> 7677.0 kB +- 219.3 kB: 1.04x faster
- scimark_lu: 8120.4 kB +- 340.0 kB -> 7843.8 kB +- 348.3 kB: 1.04x faster
- sympy_integrate: 60.9 MB +- 424.5 kB -> 58.8 MB +- 456.4 kB: 1.04x faster
- meteor_contest: 9609.0 kB +- 374.0 kB -> 9283.7 kB +- 125.0 kB: 1.04x faster
- scimark_fft: 8204.3 kB +- 217.3 kB -> 7927.2 kB +- 227.9 kB: 1.03x faster
- spectral_norm: 7379.3 kB +- 241.6 kB -> 7147.8 kB +- 271.3 kB: 1.03x faster
- nbody: 7564.3 kB +- 252.4 kB -> 7327.0 kB +- 259.3 kB: 1.03x faster
- scimark_sor: 8102.5 kB +- 196.4 kB -> 7852.7 kB +- 352.7 kB: 1.03x faster
- crypto_pyaes: 7926.6 kB +- 120.6 kB -> 7689.8 kB +- 206.4 kB: 1.03x faster
- sympy_str: 61.3 MB +- 39.1 kB -> 59.4 MB +- 51.4 kB: 1.03x faster
- richards: 7951.8 kB +- 286.5 kB -> 7715.4 kB +- 244.9 kB: 1.03x faster
- unpack_sequence: 9026.1 kB +- 296.4 kB -> 8760.5 kB +- 387.8 kB: 1.03x faster
- django_template: 37.8 MB +- 116.9 kB -> 36.7 MB +- 109.5 kB: 1.03x faster
- sympy_expand: 60.1 MB +- 44.5 kB -> 58.4 MB +- 50.4 kB: 1.03x faster
- sympy_sum: 71.9 MB +- 2189.4 kB -> 69.9 MB +- 2175.3 kB: 1.03x faster
- raytrace: 8276.0 kB +- 152.7 kB -> 8050.4 kB +- 161.2 kB: 1.03x faster
- dulwich_log: 15.3 MB +- 94.2 kB -> 14.9 MB +- 99.3 kB: 1.03x faster
- pickle_list: 7874.5 kB +- 191.8 kB -> 7666.5 kB +- 173.0 kB: 1.03x faster
- pickle: 7937.0 kB +- 203.7 kB -> 7731.4 kB +- 386.8 kB: 1.03x faster
- regex_effbot: 8133.9 kB +- 222.4 kB -> 7924.1 kB +- 237.5 kB: 1.03x faster
- unpickle_list: 7853.8 kB +- 211.0 kB -> 7656.6 kB +- 201.3 kB: 1.03x faster
- deltablue: 9946.5 kB +- 169.6 kB -> 9698.5 kB +- 98.9 kB: 1.03x faster
- sqlite_synth: 9487.7 kB +- 41.4 kB -> 9255.3 kB +- 40.3 kB: 1.03x faster
- unpickle_pure_python: 7912.4 kB +- 187.2 kB -> 7720.2 kB +- 209.8 kB: 1.02x faster
- tornado_http: 29.9 MB +- 899.9 kB -> 29.2 MB +- 831.3 kB: 1.02x faster
- scimark_sparse_mat_mult: 8554.4 kB +- 114.1 kB -> 8358.8 kB +- 340.0 kB: 1.02x faster
- go: 9354.8 kB +- 372.5 kB -> 9142.5 kB +- 367.4 kB: 1.02x faster
- xml_etree_iterparse: 12.4 MB +- 233.6 kB -> 12.1 MB +- 307.8 kB: 1.02x faster
- pathlib: 9057.6 kB +- 173.9 kB -> 8863.2 kB +- 166.3 kB: 1.02x faster
- chameleon: 20.0 MB +- 346.1 kB -> 19.6 MB +- 316.4 kB: 1.02x faster
- regex_v8: 13.2 MB +- 143.0 kB -> 12.9 MB +- 199.2 kB: 1.02x faster
- python_startup_no_site: 11.3 MB +- 34.3 kB -> 11.1 MB +- 23.3 kB: 1.02x faster
- python_startup: 11.3 MB +- 39.6 kB -> 11.1 MB +- 24.2 kB: 1.02x faster
- xml_etree_parse: 11.7 MB +- 90.7 kB -> 11.5 MB +- 464.2 kB: 1.02x faster
- 2to3: 22.6 MB +- 39.0 kB -> 22.3 MB +- 45.5 kB: 1.01x faster
- json_dumps: 9413.9 kB +- 48.8 kB -> 9294.5 kB +- 325.8 kB: 1.01x faster

Benchmark hidden because not significant (7): float, hexiom, logging_format, mako, nqueens, pyflate, scimark_monte_carlo

Geometric mean: 1.03x faster
```

</details>

Ignore `pyperf`'s incorrect "faster"/"slower" terminology... we're measuring memory usage here. I'm still waiting on actual performance numbers for this.

File: Objects/codeobject.c (hunk 11 above)
Comment (brandtbucher) [reply Objects/codeobject.c:1175]: I'd prefer to leave it as-is for now. We can always cache it later if it shows a clear improvement for a legitimate use case, or when it indeed becomes necessary for instrumentation. For the time being, I'm not sure that adding an almost-always-unused member to every code object is worth pushing 4 actual bytecode instructions out of `_co_code`'s first cache line.

(FYI, this same reasoning is why I also removed the existing caches for `co_varnames`, `co_cellvars`, and `co_freevars`.)

File: Objects/codeobject.c (hunk 3 above)
Comment (brandtbucher) [reply Objects/codeobject.c:1141]: > This no longer gets the `varnames`, it makes it.

The functionality hasn't changed, it just doesn't cache its result anymore. I'd rather just leave the name alone, since it's quite clear as-is.

> Also, why this change?

These three members are hardly used, and recomputing them is not expensive (in time *or* space). I also don't think they're worth the additional cache pressure they likely place on actual bytecode now.

File: Objects/codeobject.c (hunk 4 above)
Comment (brandtbucher) [reply Objects/codeobject.c]: It was supposed to be "underscore code" (as in `_co_code`)... but I see now that it's not a very good name at all. Haha.

File: Objects/codeobject.c (hunk 5 above)
Comment (brandtbucher) [reply Objects/codeobject.c]: Sounds good. I'll rename the `PyCodeObject` member too, for consistency.

File: Objects/codeobject.c (hunk 11 above)
Comment (markshannon) [reply Objects/codeobject.c:1175]: I agree with removing the caches for `co_varnames`, `co_cellvars`, and `co_freevars`.
We will need to add `co_code` back, but it doesn't have in this PR.

Regarding caching, it probably matters more that the first instruction is at the start of a cache line. 
But let's not worry about it here, although we will need revisit the layout, and marshaling, of code objects before 3.11b.

File: Tools/scripts/deepfreeze.py [hunk 12]
```diff
@@ -249,14 +245,44 @@ def generate_code(self, name: str, code: types.CodeType) -> str:
...
+                self.write("PyObject *co_localsplusnames;")
+                self.write("PyObject *co_localspluskinds;")
+                self.write("PyObject *co_filename;")
+                self.write("PyObject *co_name;")
+                self.write("PyObject *co_qualname;")
+                self.write("PyObject *co_linetable;")
+                self.write("PyObject *co_endlinetable;")
+                self.write("PyObject *co_columntable;")
+                self.write("PyObject *co_weakreflist;")
+                self.write("void *co_extra;")
```
Comment (gvanrossum) [inline Tools/scripts/deepfreeze.py]: I think this should be a definition in code.h so you don't have to update this every time the code object changes.

**Comment by brandtbucher** [comment]
1% perf improvement too:

<details>

```
Slower (11):
- pickle_dict: 27.6 us +- 0.2 us -> 28.4 us +- 0.3 us: 1.03x slower
- html5lib: 65.3 ms +- 2.7 ms -> 67.2 ms +- 2.8 ms: 1.03x slower
- pickle_list: 4.33 us +- 0.05 us -> 4.46 us +- 0.05 us: 1.03x slower
- regex_v8: 23.1 ms +- 0.2 ms -> 23.8 ms +- 0.2 ms: 1.03x slower
- regex_dna: 217 ms +- 1 ms -> 223 ms +- 4 ms: 1.03x slower
- scimark_lu: 111 ms +- 2 ms -> 113 ms +- 2 ms: 1.02x slower
- regex_effbot: 3.46 ms +- 0.06 ms -> 3.50 ms +- 0.05 ms: 1.01x slower
- json_dumps: 12.6 ms +- 0.1 ms -> 12.8 ms +- 0.2 ms: 1.01x slower
- fannkuch: 397 ms +- 3 ms -> 400 ms +- 5 ms: 1.01x slower
- json_loads: 28.1 us +- 0.3 us -> 28.3 us +- 0.3 us: 1.01x slower
- xml_etree_iterparse: 105 ms +- 1 ms -> 105 ms +- 1 ms: 1.01x slower

Faster (43):
- go: 149 ms +- 1 ms -> 139 ms +- 1 ms: 1.07x faster
- logging_simple: 5.38 us +- 0.10 us -> 5.16 us +- 0.08 us: 1.04x faster
- pickle: 9.89 us +- 0.14 us -> 9.53 us +- 0.10 us: 1.04x faster
- pycparser: 1.24 sec +- 0.02 sec -> 1.19 sec +- 0.02 sec: 1.04x faster
- thrift: 780 us +- 13 us -> 754 us +- 8 us: 1.03x faster
- deltablue: 3.85 ms +- 0.05 ms -> 3.73 ms +- 0.05 ms: 1.03x faster
- unpack_sequence: 48.5 ns +- 0.5 ns -> 47.1 ns +- 0.9 ns: 1.03x faster
- scimark_sparse_mat_mult: 4.97 ms +- 0.15 ms -> 4.83 ms +- 0.11 ms: 1.03x faster
- pyflate: 451 ms +- 3 ms -> 438 ms +- 4 ms: 1.03x faster
- xml_etree_process: 56.7 ms +- 0.8 ms -> 55.2 ms +- 0.7 ms: 1.03x faster
- pickle_pure_python: 323 us +- 3 us -> 314 us +- 3 us: 1.03x faster
- telco: 6.80 ms +- 0.09 ms -> 6.65 ms +- 0.16 ms: 1.02x faster
- scimark_sor: 120 ms +- 1 ms -> 118 ms +- 1 ms: 1.02x faster
- pidigits: 194 ms +- 0 ms -> 190 ms +- 0 ms: 1.02x faster
- logging_format: 5.87 us +- 0.08 us -> 5.74 us +- 0.09 us: 1.02x faster
- unpickle_pure_python: 238 us +- 2 us -> 233 us +- 2 us: 1.02x faster
- xml_etree_generate: 80.0 ms +- 0.6 ms -> 78.4 ms +- 0.7 ms: 1.02x faster
- meteor_contest: 108 ms +- 3 ms -> 106 ms +- 1 ms: 1.02x faster
- regex_compile: 139 ms +- 1 ms -> 136 ms +- 1 ms: 1.02x faster
- hexiom: 6.96 ms +- 0.03 ms -> 6.83 ms +- 0.02 ms: 1.02x faster
- sympy_sum: 163 ms +- 2 ms -> 160 ms +- 1 ms: 1.02x faster
- tornado_http: 98.2 ms +- 1.3 ms -> 96.5 ms +- 1.4 ms: 1.02x faster
- dulwich_log: 65.8 ms +- 0.4 ms -> 64.7 ms +- 0.5 ms: 1.02x faster
- sympy_integrate: 20.9 ms +- 0.1 ms -> 20.6 ms +- 0.1 ms: 1.02x faster
- scimark_fft: 340 ms +- 4 ms -> 334 ms +- 4 ms: 1.02x faster
- 2to3: 267 ms +- 1 ms -> 263 ms +- 1 ms: 1.02x faster
- scimark_monte_carlo: 69.7 ms +- 1.2 ms -> 68.7 ms +- 0.8 ms: 1.01x faster
- django_template: 35.0 ms +- 0.5 ms -> 34.5 ms +- 0.5 ms: 1.01x faster
- chaos: 71.7 ms +- 0.6 ms -> 70.7 ms +- 0.6 ms: 1.01x faster
- nbody: 94.0 ms +- 1.7 ms -> 92.8 ms +- 1.8 ms: 1.01x faster
- raytrace: 310 ms +- 2 ms -> 306 ms +- 3 ms: 1.01x faster
- sqlalchemy_declarative: 141 ms +- 3 ms -> 140 ms +- 3 ms: 1.01x faster
- float: 76.7 ms +- 0.8 ms -> 75.8 ms +- 1.0 ms: 1.01x faster
- sympy_str: 291 ms +- 2 ms -> 287 ms +- 3 ms: 1.01x faster
- richards: 47.5 ms +- 1.2 ms -> 47.0 ms +- 1.1 ms: 1.01x faster
- sympy_expand: 485 ms +- 6 ms -> 480 ms +- 3 ms: 1.01x faster
- python_startup_no_site: 6.02 ms +- 0.00 ms -> 5.96 ms +- 0.00 ms: 1.01x faster
- chameleon: 6.63 ms +- 0.07 ms -> 6.57 ms +- 0.06 ms: 1.01x faster
- crypto_pyaes: 83.9 ms +- 0.7 ms -> 83.2 ms +- 1.1 ms: 1.01x faster
- spectral_norm: 102 ms +- 1 ms -> 101 ms +- 1 ms: 1.01x faster
- python_startup: 8.41 ms +- 0.01 ms -> 8.34 ms +- 0.01 ms: 1.01x faster
- nqueens: 86.1 ms +- 1.2 ms -> 85.5 ms +- 0.8 ms: 1.01x faster
- pathlib: 18.3 ms +- 0.2 ms -> 18.2 ms +- 0.3 ms: 1.01x faster

Benchmark hidden because not significant (8): json, logging_silent, mako, sqlalchemy_imperative, sqlite_synth, unpickle, unpickle_list, xml_etree_parse

Geometric mean: 1.01x faster
```

</details>

File: Include/cpython/code.h [hunk 13]
```diff
@@ -26,75 +26,80 @@ typedef uint16_t _Py_CODEUNIT;
...
+                                     lines */                                  \
+    PyObject *co_columntable;     /* bytes object that holds start/end column  \
+                                     offset each instruction */                \
+                                                                               \
+    PyObject *co_weakreflist;     /* to support weakrefs to code objects */    \
+    /* Scratch space for extra data relating to the code object.               \
+       Type is a void* to keep the format private in codeobject.c to force     \
+       people to go through the proper APIs. */                                \
+    void *co_extra;                                                            \
+    char co_code_adaptive[(SIZE)];                                             \
```
Comment (gvanrossum) [inline Include/cpython/code.h:98]: Nice!

Comment (markshannon) [reply Include/cpython/code.h:98]: Not a fan of giant macros, myself. They make things harder to parse, both for tools and me.

Maybe something like
```
typedef struct _code_object_fields {
    ... 
} _code_object_fields;

struct PyCodeObject {
    PyObject_VAR_HEAD 
    _code_object_fields body;
    char co_code_adaptive[1];
}
```
Then in the generated code:
```
#define _PyCode_DEF(SIZE) { \
    PyObject_VAR_HEAD  \
    _code_object_fields body; \
    char co_code_adaptive[(SIZE)];    \
}
```
Which avoids most of the code duplication.

**Comment by markshannon** [comment]
This is still marked as draft, what is left to do?

File: Include/cpython/code.h (hunk 13 above)
Comment (markshannon) [reply Include/cpython/code.h:98]: This does add some extra code though, `code->co_xxx` becomes something like `code->co_body.xxx`.
It's up to you.

**Comment by brandtbucher** [comment]
> This is still marked as draft, what is left to do?

There is still an awkward spot in `_gen_throw` where we walk back `f_lasti` to the previous `SEND` instruction and perform the jump ourselves when in a `yield from` (which is sort of a strange control-flow path that isn’t reflected in the CFG/bytecode/`dis`). I also don’t think the current implementation handles `EXTENDED_ARG`s correctly.

I’m still trying to understand the code better and figure out a cleaner way of doing this. Any ideas?

File: Include/cpython/code.h (hunk 13 above)
Comment (brandtbucher) [reply Include/cpython/code.h:98]: I'd prefer not to go through the intermediate struct.

I'm also not a *huge* fan of the big macro, but it does significantly reduce the duplication across these files (and the possibility of hard-to-debug errors when modifying only one of them).

**Comment by brandtbucher** [comment]
> There is still an awkward spot in `_gen_throw` where we walk back `f_lasti` to the previous `SEND` instruction and perform the jump ourselves when in a `yield from` (which is sort of a strange control-flow path that isn’t reflected in the CFG/bytecode/`dis`). I also don’t think the current implementation handles `EXTENDED_ARG`s correctly.
> 
> I’m still trying to understand the code better and figure out a cleaner way of doing this. Any ideas?

https://github.com/python/cpython/pull/31968 should help.

**Comment by markshannon** [comment]
I think it is OK to sort out the peculiarities of `gen.throw()` after this PR is merged.

File: Include/cpython/code.h (hunk 13 above)
Comment (markshannon) [reply Include/cpython/code.h:98]: Ok, let's keep the macro.

File: Objects/codeobject.c [hunk 14]
```diff
@@ -1348,25 +1330,21 @@ code_dealloc(PyCodeObject *co)
...
     Py_XDECREF(co->co_columntable);
     Py_XDECREF(co->co_exceptiontable);
-    if (co->co_weakreflist != NULL)
+    if (co->co_weakreflist != NULL) {
         PyObject_ClearWeakRefs((PyObject*)co);
-    if (co->co_quickened) {
-        PyMem_Free(co->co_quickened);
+    }
+    if (co->co_warmup == 0) {
         _Py_QuickenedCount--;
```
Comment (markshannon) [inline Objects/codeobject.c:1348]: I think we can remove `_Py_QuickenedCount` now. IIRC, it exists only to account for extra allocation and stop the refleak checker from complaining.

No need to block this PR though, and it might be a useful counter to have anyway.
If we do keep it, we should probably move it to the stats.

Comment (markshannon) [reply Objects/codeobject.c:1348]: Added to https://github.com/faster-cpython/ideas/issues/310

File: Objects/codeobject.c [hunk 15]
```diff
@@ -1420,8 +1398,21 @@ code_richcompare(PyObject *self, PyObject *other, int op)
     if (!eq) goto unequal;
     eq = co->co_firstlineno == cp->co_firstlineno;
     if (!eq) goto unequal;
-    eq = PyObject_RichCompareBool(co->co_code, cp->co_code, Py_EQ);
-    if (eq <= 0) goto unequal;
+    PyObject *co_code = _PyCode_GetCode(co);
```
Comment (markshannon) [inline Objects/codeobject.c:1401]: No need to create both code objects here. Iterating over both bytecode arrays, comparing `_PyOpcode_Deopt[op]` and skipping caches would be considerably more efficient.

Added to https://github.com/faster-cpython/ideas/issues/310

**Comment by markshannon** [comment]
Buildbot failures are all pre-existing failures and unrelated to this PR.
